serial/serbridge.c

```diff
@@ -25,117 +61,523 @@ void (*programmingCB)(char *buffer, short length) = NULL;
+    else
+      uart0_write_char(c) ;
+    return ;
+  }
+
+  if ( state == ST_IAC ) {
+    if ( c == IAC ) {
+      /*
+       *  Dual IAC means char \xFF
+       */
+      uart0_write_char(0xFF);
+      state = ST_NORMAL ;
+      return ;
+    }
+    else if ( c == DONT || c == DO || c == WONT || c == WILL ) {
+      /*
+       *  Beginning of negociation
+       */
+      opt = c ;
+      state = ST_NEGO ;
+      return ;
+    }
+    else if ( c == SB ) {
+      /*
+       *  Beginning of sub option
+       */
+      opt = c ;
+      state = ST_SB ;
+      return ;
+    }
+    else if ( c == SE ) {
+      /*
+       *  End of sub option
+       */
+      opt = c ;
+      state = ST_NORMAL ;
+      return ;
+    }
+    return ;
+  }
+
+  if ( state == ST_NEGO ) {
+    dbgf("TELNET: IAC NEGO (%02X)", c);
+    if ( c == BINARY ||
+	 c == ECHO ||
+	 c == SUPPRESS_GO_AHEAD ||
+	 c == COM_PORT_OPTION ) {
+      /*
+       *  Acknowledge positive requests for known options
+       */
+      dbgf(": OK\n");
+      if ( opt == DO )
+	opt = WILL ;
+      else if ( opt == WILL )
+	opt = DO ;
+    }
+    else {
+      /*
+       *  Deny positive requests for unknown options
+       */
+      dbgf(": REJECTED\n");
+      if ( opt == DO )
+	opt = WONT ;
+      else if ( opt == WILL )
+	opt = DONT ;
+    }
+    /*
+     *  Send reply
+     */
+    espbuffsend( conn, (char[]){IAC,opt,c}, 3 );
+    state = ST_NORMAL ;
+    return ;
+  }
+
+  if ( state == ST_SB ) {
+    if ( c == COM_PORT_OPTION )
+      state = ST_COMPORT ;
+    else
+      state = ST_WAIT_IAC ;
+    return ;
+  }
+
+  if ( state == ST_WAIT_IAC ) {
+    if ( c == IAC )
+      state = ST_IAC ;
+    return ;
+  }
+
+  if ( state == ST_COMPORT ) {
+    if ( c == SET_BAUDRATE ) {
+      dbgf("  BAUDRATE:");
+      vlen = 0 ;
+      state = ST_BAUDRATE ;
+    } else if ( c == SET_DATASIZE ) {
+      dbgf("  DATASIZE:");
+      state = ST_DATASIZE ;
+    } else if ( c == SET_PARITY ) {
+      dbgf("  PARITY:");
+      state = ST_PARITY ;
+    } else if ( c == SET_STOPSIZE ) {
+      dbgf("  STOPSIZE:");
+      state = ST_STOPSIZE ;
+    } else if ( c == SET_CONTROL ) {
+      dbgf("  CONTROL:");
+      state = ST_CONTROL ;
+    } else if ( c == PURGE_DATA ) {
+      dbgf("  PURGE:");
+      state = ST_PURGE ;
+    } else {
+      dbgf("UNKNOWN: %02X\n", c);
+      state = ST_WAIT_IAC ;
+    }
+    return ;
+  }
+
+  if ( state == ST_BAUDRATE ) {
+    /*
+     *  Get 4 bytes of baudrate (MSB first)
+     */
+    dbgf(" %02X", c);
+
+    value <<= 8 ;
+    value += c ;
+    vlen++ ;
+    if ( vlen == 4 ) {
+      /*
+       *  4 bytes received, process the value
+       *
+       *  Note: must acknowledge with the value that was set otherwise the
+       *  pyserial client considers that the set value is rejected.
+       */
+      if ( value == 0 )
+	/*
+	 *  Get actual baudrate
+	 */
+	value = (int)(0.5 + 80e6/REG_BRR);
+      else
+	/*
+	 *  Set baudrate
+	 */
+	REG_BRR = (int)(0.5 + 80e6/value);
```

Comment:
This makes it completely decoupled from the uart driver and from the settings in flash, causing the web interface not to show the correct value either...